ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/util/ManifestUtils.java

```diff
@@ -285,4 +303,96 @@ private static String splitOnComma(String value) {
 		sb.append(values[values.length - 1]);
 		return sb.toString();
 	}
+
+	/**
+	 * Returns the list all execution-environments required by the given OSGi
+	 * resource. Only registered EEs are considered.
+	 *
+	 * @param resource
+	 *            the osgi resource
+	 * @return a list containing the id's of all required EEs
+	 * @see ExecutionEnvironmentNamespace#EXECUTION_ENVIRONMENT_NAMESPACE
+	 * @see IExecutionEnvironmentsManager#getExecutionEnvironments()
+	 */
+	public static List<String> getRequiredExecutionEnvironments(Resource resource) {
+		if (resource instanceof BundleDescription bundleDescription) {
+			// fast path if a BREE header is specified in the MANIFEST.MF
+			String[] requiredEE = bundleDescription.getExecutionEnvironments();
+			if (requiredEE.length > 0) {
+				return List.of(requiredEE);
```

Comment:
https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module.bree

> This header is deprecated but must be fully supported by a compliant framework. Bundles should not mix these headers but use either an osgi.ee requirement or this header. **If both are used, both constraints must be met to resolve**. 

If one would be pedantic then always both headers have to be considered.

Reply:
Yes and I thought about that as well. I mainly added it as a short-cut/fast-path for bundles that define a BREE. And although you are right that both have to be considered I have never seen a bundle that really has both.

But when PDE fully migrates off the Equinox resolver API that section has to be removed without replacement anyway, so it could just not be added in the first place.

Reply:
Lets keep it simple and correct. Just removed that premature optimization. This also avoids adjustments in the future if PDE moves off the Equinox resolver.